server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -347,4 +361,13 @@ private static UnassignedInfo discardAllocationStatus(UnassignedInfo info) {
             info.getLastAllocatedNodeId()
         );
     }
+
+    private static int computeIterationCountWarningInterval(RoutingAllocation allocation) {
+        final int relativeSize = allocation.metadata().getTotalNumberOfShards();
+        int iterations = 1000;
+        while (iterations < relativeSize) {
```

Comment:
Makes sense. Slight risk of overflow here if we were to break the billion-shard barrier tho - could we be explicit about that just to be safe?

```suggestion
        while (iterations < relativeSize && iterations < 1_000_000_000) {
```